scripts/research/lib/research_harness.py

```
@@ -0,0 +1,940 @@
+from __future__ import annotations
+
+import hashlib
+import json
+import math
+import os
+import subprocess
+from dataclasses import dataclass, field
+from datetime import datetime, timezone
+from numbers import Integral
+from pathlib import Path
+from typing import AbstractSet, Hashable, Mapping, Sequence
```

Ruff (0.16.2)

[warning] Import from `collections.abc` instead: `Hashable`, `Mapping`, `Sequence`

Import from `collections.abc`

(UP035)